packages/dev/s2-docs/src/client.tsx

```diff
@@ -15,8 +19,53 @@ let updateRoot = hydrate({
 // and in a React transition, stream in the new page. Once complete, we'll pushState to
 // update the URL in the browser.
 async function navigate(pathname: string, push = false) {
+  let loadingShown = false;
+  let [basePath] = pathname.split('#');
+  let rscPath = basePath.replace('.html', '.rsc');
+  
+  // Use prefetched result if available, otherwise fetch
+  const prefetchedPromise = getPrefetchedPromise(rscPath);
+  const fetchPromise = prefetchedPromise ?? fetchRSC<ReactElement>(rscPath);
+  
+  const delayPromise = new Promise<void>(resolve => setTimeout(resolve, 50));
+  
+  // Race the fetch against a small delay to detect cached responses
+  // If fetch resolves before the delay, we won't show the skeleton
+  let raceResult: 'fast' | 'slow' | 'error' = 'slow';
   try {
-    let res = await fetchRSC<ReactElement>(pathname.replace('.html', '.rsc'));
+    raceResult = await Promise.race([
+      fetchPromise.then(() => 'fast' as const).catch(() => 'error' as const),
+      delayPromise.then(() => 'slow' as const)
```

Comment:
We may not need the delay - we kinda want the skeleton to show up as fast as possible so the user knows something is happening, but just not when the result is already cached.